_posts/2020-04-17-quarkus-mutual-tls.adoc

```diff
@@ -0,0 +1,405 @@
+---
+layout: post
+title: 'Quarkus - Makes Mutual TLS Easy.'
+date:   2020-04-17 00:00 +0100
+author: mkouba
+synopsis: Quarkus - Mutual TLS Authentication.
+---
+
+== What is Mutual TLS Authentication?
+
+Mutual TLS authentication or two way authentication is an extension of Transport layer security (or “TLS”), and it ensures that traffic between the client and server is secure and trusted in both directions.
+
+== Why Mutual TLS?
+
+It is an additional security layer which verifies the identity of the client.
+
+== How implement it?
+
+The best approach to implement Mutual TLS between two services has to come from the infrastructure, for instance, Service Mesh, to achieve a standard and secure way of communication and to avoid that each application implements their own solution. However, you are not always working on cutting edge environment.
+
+Let's have a look at how to implement mTLS with Quarkus.
+ 
+== Bootstrapping.
+
+Bootstrapping Hello server application,
+
+[source]
+----
+mvn io.quarkus:quarkus-maven-plugin:1.4.0.CR1:create \
+    -DprojectGroupId=org.acme \
+    -DprojectArtifactId=quarkus-server-mtls \
+    -DclassName="org.acme.server.mtls.GreetingResource" \
+    -Dpath="/hello" \
+    -DplatformArtifactId=quarkus-bom \
+    -DplatformVersion=1.4.0.CR1
+----
+
+and Hello client application:
+
+[source]
+----
+mvn io.quarkus:quarkus-maven-plugin:1.4.0.CR1:create \
+    -DprojectGroupId=org.acme \
+    -DprojectArtifactId=quarkus-client-mtls \
+    -DclassName="org.acme.client.mtls.GreetingResource" \
+    -Dextensions="rest-client, resteasy-jsonb" \
+    -Dpath="/hello" \
+    -DplatformArtifactId=quarkus-bom \
+    -DplatformVersion=1.4.0.CR1
+----
+
+== Certificate and Truststore generation.
+
+Of course, you need a server, client certificates and a truststore :)
+
+[source]
+----
+keytool -genkeypair -storepass password -keyalg RSA -keysize 2048 -dname "CN=server" -alias server -ext "SAN:c=DNS:localhost,IP:127.0.0.1" -keystore quarkus-server-mtls/src/main/resources/META-INF/resources/server.keystore
+
+keytool -genkeypair -storepass password -keyalg RSA -keysize 2048 -dname "CN=client" -alias client -ext "SAN:c=DNS:localhost,IP:127.0.0.1" -keystore quarkus-client-mtls/src/main/resources/META-INF/resources/client.keystore
+----
+
+NOTE: For this example, the `client.keystore` used as _truststore_ for the server application and in the same way the `server.keystore` for the client application.
+
+[source]
```

Comment:
it would be good to explain why you are doing that

Reply:
To use the certificates as truststore, I wanted to make it simple without explaining additional steps, certificate authority extract the certificate, etc etc